plugins/delete-by-query/src/main/java/org/elasticsearch/deletebyquery/action/TransportDeleteByQueryAction.java

```diff
@@ -0,0 +1,454 @@
+                        executeScroll(searchResponse.getScrollId());
+                    }
+
+                    @Override
+                    public void onFailure(Throwable e) {
+                        listener.onFailure(e);
+                    }
+                });
+            } catch (Throwable t) {
+                logger.error("unable to execute the initial scan request of delete by query", t);
+                listener.onFailure(t);
+            }
+        }
+
+        void executeScroll(final String scrollId) {
+            threadPool.generic().execute(new AbstractRunnable() {
+                @Override
+                protected void doRun() throws Exception {
+                    if (logger.isDebugEnabled()) {
+                        logger.debug("executing scroll request [{}]", scrollId);
+                    }
+
+                    scrollAction.execute(new SearchScrollRequest(scrollId).scroll(request.scroll()), new ActionListener<SearchResponse>() {
+                        @Override
+                        public void onResponse(SearchResponse scrollResponse) {
+                            onScrollResponse(scrollId, scrollResponse);
+                        }
+
+                        @Override
+                        public void onFailure(Throwable e) {
+                            onScrollFailure(scrollId, e);
+
+                        }
+                    });
+                }
+
+                @Override
+                public void onFailure(Throwable t) {
+                    onScrollFailure(scrollId, t);
+                }
+
+                @Override
+                public void onRejection(Throwable t) {
+                    onScrollFailure(scrollId, t);
+                }
+            });
+        }
+
+        void onScrollResponse(String scrollId, SearchResponse scrollResponse) {
+            final SearchHit[] docs = scrollResponse.getHits().getHits();
+            final String nextScrollId = scrollResponse.getScrollId();
+            addShardFailures(scrollResponse.getShardFailures());
+
+            if (logger.isDebugEnabled()) {
+                logger.debug("scroll request [{}] executed: [{}] document(s) returned", scrollId, docs.length);
+            }
+
+            if ((docs.length == 0) || (nextScrollId == null)) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("scrolling documents terminated");
+                }
+                finishScroll(scrollId, false, null);
+                return;
+            }
+
+            if (isTimedOut()) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("scrolling documents timed out");
+                }
+                finishScroll(scrollId, true, null);
+                return;
+            }
+
+            // Delete the scrolled documents using the Bulk API
+            threadPool.generic().execute(new AbstractRunnable() {
```

Comment:
I don't think we need to execute it on a thread pool, nothing blocking, we can do it without one and delegate to the async bulk execution